Kubernetes Single Step APM Instrumentation (#20573)
DataDog/documentation: content/en/tracing/trace_collection/single-step-apm.md

     -v /var/run/docker.sock:/var/run/docker.sock:ro \
     gcr.io/datadoghq/agent:7
   ```

>    Replace `<YOUR_DD_API_KEY>` with your [Datadog API][4].

Reviewer: This link is broken (I think you just need to add it to this tab)

3. Restart the Docker containers.
4. [Explore the performance observability of your services in Datadog][2].

   <div class="alert alert-info">You can optionally set an environment for your services and other telemetry that pass through the Agent. Read <a href="#tagging-observability-data-by-environment">tagging observability data by environment</a> to learn how. </div>

[1]: /agent/remote_config
[2]: /tracing/service_catalog/

{{% /tab %}}

{{% tab "Kubernetes" %}}

You can enable APM when installing the Agent with the [Datadog Operator](#operator) or [Helm chart](#helm-chart).

### Operator

The Datadog Operator is a way to deploy the Datadog Agent on Kubernetes and OpenShift.

#### Prerequisites

> - Kubernetes Cluster version v1.20.X+: Tests were done on v1.20.0+; should be supported in v1.11.0+. For earlier versions, because of limited CRD support, the Operator may not work as expected.

Reviewer: ```suggestion
- Kubernetes Cluster version v1.20.X+ and v1.11.0+. For earlier versions, because of limited CRD support, the Operator may not work as expected.
```
This construction is a little confusing and I think we should avoid "should". Also, I wonder whether the versions that were tested need to be included? If we say a range is supported, the user can probably infer that we've tested that range.

- [`Helm`][2] for deploying the `datadog-operator`.

> - [`Kubectl` CLI][3] for installing the `datadog-agent`.

Reviewer: This currently links to the datadog API docs

#### Installation

To enable instrumentation with the Datadog Operator:

1. Install the [Datadog Operator][5]:

   ```shell
   helm repo add datadog https://helm.datadoghq.com
   helm install my-datadog-operator datadog/datadog-operator
   ```

2. Create a Kubernetes secret with your API and app keys

   ```shell
   kubectl create secret generic datadog-secret --from-literal api-key=<DATADOG_API_KEY> --from-literal app-key=<DATADOG_APP_KEY>
   ```

>    Replace `<DATADOG_API_KEY>` and `<DATADOG_APP_KEY>` with your [Datadog API][6] and [application keys][7].

Reviewer: The Datadog API one links to K8s log collection. Not sure if that's intended

2. Create a file, `datadog-agent.yaml`, with the spec of your Datadog Agent deployment configuration. The simplest configuration is as follows:

   ```yaml
   kind: DatadogAgent
   apiVersion: datadoghq.com/v2alpha1
   metadata:
     name: datadog
   spec:
     global:
       site: <DATADOG_SITE>
       credentials:
         apiSecret:
           secretName: datadog-secret
           keyName: api-key
         appSecret:
           secretName: datadog-secret
           keyName: app-key
   features:
     apm:
       instrumentation:
         enabled: true  
     override:
       clusterAgent:
         image:
           name: gcr.io/datadoghq/cluster-agent:latest
       nodeAgent:
         image:
           name: gcr.io/datadoghq/agent:latest
   ```

>    Replace `<DATADOG_SITE>` with your [Datadog site][10]. Your site is {{< region-param key="dd_site" code="true" >}}. (Ensure the correct **DATADOG SITE** is selected on this page).

Reviewer: Datadog site links to the Docker docs

   helm install <RELEASE_NAME> \
    -f datadog-values.yaml \
    --set targetSystem=<TARGET_SYSTEM> \
    datadog/datadog
   ```

- `<RELEASE_NAME>`: Your release name. For example, `datadog-agent`.
- `<TARGET_SYSTEM>`: The name of your OS. For example, `linux` or `windows`.

**Note**: If you are using Helm `2.x`, run the following:
   ```bash
   helm install --name <RELEASE_NAME> \
    -f datadog-values.yaml \
    --set targetSystem=<TARGET_SYSTEM> \
    datadog/datadog
   ```

[1]: https://v3.helm.sh/docs/intro/install/
[2]: https://github.com/DataDog/helm-charts/blob/master/charts/datadog/values.yaml
[3]: https://app.datadoghq.com/organization-settings/api-keys
[4]: https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-state-metrics
[5]: /agent/kubernetes/apm?tab=helm
[6]: /agent/kubernetes/log?tab=helm
[8]: https://gcr.io/datadoghq
[9]: https://gallery.ecr.aws/datadog/
[10]: https://hub.docker.com/u/datadog/
[11]: https://github.com/DataDog/helm-charts/blob/master/charts/datadog/docs/Migration_1.x_to_2.x.md
[12]: /integrations/kubernetes_state_core
[13]: /getting_started/site
[14]: https://github.com/DataDog/helm-charts/blob/main/charts/datadog/values.yaml

> [15]: https://app.datadoghq.com/organization-settings/application-keys

Reviewer: Curious if the tab was meant to end here after the links

Author: Good question--the tab is supposed to include enabling or disabling namespaces + specifying library versions, because those are only supported for Kubernetes.

### Enabling or disabling namespaces

You can choose to selectively instrument or to not instrument specific namespaces.

#### Operator

To enable instrumentation for specific namespaces:

1. Add the `enabledNamespaces` configuration to your `datadog-agent.yaml` file:

> {{< highlight yaml "hl_lines=19-21" >}}

Reviewer: You can probably indent this first line to bring it in line with the list.

Author: Yeah, the highlight shortcode doesn't like indenting, but I agree it looks messy. Changed to normal code fence, so I could indent properly.

   kind: DatadogAgent
   apiVersion: datadoghq.com/v2alpha1
   metadata:
     name: datadog
   spec:
     global:
       site: <DATADOG_SITE>
       credentials:
         apiSecret:
           secretName: datadog-secret
           keyName: api-key
         appSecret:
           secretName: datadog-secret
           keyName: app-key
   features:
     apm:
       instrumentation:
         enabled: true 
         enabledNamespaces:
           - default
           - applications
     override:
       clusterAgent:
         image:
           name: gcr.io/datadoghq/cluster-agent:latest
       nodeAgent:
         image:
           name: gcr.io/datadoghq/agent:latest
{{< /highlight >}}

2. Restart your services.

To disable instrumentation for specific namespaces:

1. Add the `disabledNamespaces` configuration to your `datadog-agent.yaml` file:

> {{< highlight yaml "hl_lines=19-21" >}}

Reviewer: ditto on the indent here and below

{{% /tab %}}

{{% tab "Kubernetes" %}}

1. Under `apm:`, remove `instrumentation:` and all following configuration in `datadog-values.yaml`.
2. Run the following command:
   ```bash

>    helm upgrade <RELEASE_NAME> -f datadog-values.yaml datadog/datadog

Author: ```suggestion
   helm upgrade datadog-agent -f datadog-values.yaml datadog/datadog
```
